File: app/src/main/java/de/westnordost/streetcomplete/quests/postbox_collection_times/CollectionTimesAdapter.java
package de.westnordost.streetcomplete.quests.postbox_collection_times;

import android.annotation.SuppressLint;
import android.app.TimePickerDialog;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;

import de.westnordost.streetcomplete.R;
import de.westnordost.streetcomplete.data.meta.CountryInfo;
import de.westnordost.streetcomplete.quests.opening_hours.model.Weekdays;
import de.westnordost.streetcomplete.quests.opening_hours.WeekdaysPickerDialog;

public class CollectionTimesAdapter extends RecyclerView.Adapter<CollectionTimesAdapter.WeekdayViewHolder>
{
	private ArrayList<WeekdaysTimes> data;
	private final Context context;
	private final CountryInfo countryInfo;

	public CollectionTimesAdapter(ArrayList<WeekdaysTimes> data, Context context, CountryInfo countryInfo)
	{
		this.data = data;
		this.context = context;
		this.countryInfo = countryInfo;
	}

	public String toString()
	{
		StringBuilder result = new StringBuilder();
		boolean firstDays = true;

		Weekdays lastWeekdays = null;
		for (WeekdaysTimes times : data)
		{
			boolean isSameWeekdays = lastWeekdays != null && lastWeekdays.equals(times.weekdays);
			if(!isSameWeekdays)
			{
				if(!firstDays)	result.append(", ");
				else            firstDays = false;

				result.append(times.weekdays.toString());
				result.append(" ");
			}
			else
			{
				result.append(",");
			}
			result.append(timeOfDayToString(times.minutes));

			lastWeekdays = times.weekdays;
		}

		return result.toString();
	}

	@Override public WeekdayViewHolder onCreateViewHolder(ViewGroup parent, int viewType)
	{
		LayoutInflater inflater = LayoutInflater.from(parent.getContext());
		return new WeekdayViewHolder(
				inflater.inflate(R.layout.quest_times_weekday_row, parent, false));
	}

	@Override public void onBindViewHolder(WeekdayViewHolder holder, int position)
	{
		WeekdaysTimes times = data.get(position);
		WeekdaysTimes previousTimes = null;
		if(position > 0) previousTimes = data.get(position - 1);
		holder.update(times, previousTimes);
	}

	@Override public int getItemCount()
	{
		return data.size();
	}

	/* ------------------------------------------------------------------------------------------ */

	private void remove(int position)
	{
		data.remove(position);
		notifyItemRemoved(position);
		// if not last weekday removed -> element after this one may need to be updated
		// because it may need to show the weekdays now
		if(position < data.size()) notifyItemChanged(position);
	}

	public void addNew()
	{
		boolean isFirst = data.isEmpty();
		openSetWeekdaysDialog(getWeekdaysSuggestion(isFirst), (weekdays) ->
		{
			openSetTimeDialog(12 * 60,	(minutes) -> add(weekdays, minutes));
		});
	}

	private void add(Weekdays weekdays, int minutes)
	{
		int insertIndex = getItemCount();
		data.add(new WeekdaysTimes(weekdays, minutes));
		notifyItemInserted(insertIndex);
	}

	public ArrayList<WeekdaysTimes> getData()
	{
		return data;
	}

	/* ------------------------------------ weekdays select --------------------------------------*/

	 class WeekdayViewHolder extends RecyclerView.ViewHolder
	{
		private TextView weekdaysText;
		private TextView hoursText;
		private View delete;

		public WeekdayViewHolder(View itemView)
		{
			super(itemView);
			weekdaysText = itemView.findViewById(R.id.weekday);
			hoursText = itemView.findViewById(R.id.hours);
			delete = itemView.findViewById(R.id.delete);
			delete.setOnClickListener(v ->
			{
				int index = getAdapterPosition();
				if(index != RecyclerView.NO_POSITION) remove(getAdapterPosition());
			});
		}

		public void update(final WeekdaysTimes times, final WeekdaysTimes previousTimes)
		{
			if(previousTimes != null && times.weekdays.equals(previousTimes.weekdays))
			{
				weekdaysText.setText("");
			}
			else
			{
				weekdaysText.setText(times.weekdays.toLocalizedString(context.getResources()));
			}

			weekdaysText.setOnClickListener(v ->
			{
				openSetWeekdaysDialog(times.weekdays, weekdays ->
				{
					times.weekdays = weekdays;
					notifyItemChanged(getAdapterPosition());
				});
			});
			hoursText.setText(timeOfDayToString(times.minutes));
			hoursText.setOnClickListener(v ->
			{
				openSetTimeDialog(times.minutes, minutes ->
				{
					times.minutes = minutes;
					notifyItemChanged(getAdapterPosition());
				});
			});
		}
	}

	@SuppressLint("DefaultLocale") private static String timeOfDayToString(int minutes)
	{
		return String.format("%02d:%02d", minutes / 60, minutes % 60);
	}

	private @NonNull Weekdays getWeekdaysSuggestion(boolean isFirst)
	{
		if(isFirst)
		{
			int firstWorkDayIdx = Weekdays.getWeekdayIndex(countryInfo.getFirstDayOfWorkweek());
			boolean[] result = new boolean[7];
			for(int i = 0; i < countryInfo.getRegularShoppingDays(); ++i)
			{
				result[(i + firstWorkDayIdx) % 7] = true;
			}
			return new Weekdays(result);
		}
		return new Weekdays();
	}

	private void openSetWeekdaysDialog(Weekdays weekdays, WeekdaysPickerDialog.OnWeekdaysPickedListener callback)
	{
		WeekdaysPickerDialog.show(context, weekdays, callback);
	}

	private interface TimePickedListener { void onTimePicked(int minutes); }

	private void openSetTimeDialog(int minutes, TimePickedListener callback)
	{
		new TimePickerDialog(context, R.style.Theme_Bubble_Dialog,
			(view, hourOfDay, minute) -> callback.onTimePicked(hourOfDay * 60 + minute),
			minutes / 60, minutes % 60,	true).show();
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/postbox_collection_times/CollectionTimesAdapter.kt
package de.westnordost.streetcomplete.quests.postbox_collection_times

import android.content.Context
import android.text.format.DateFormat
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.view.isInvisible
import androidx.recyclerview.widget.RecyclerView
import de.westnordost.streetcomplete.data.meta.CountryInfo
import de.westnordost.streetcomplete.databinding.QuestTimesWeekdayRowBinding
import de.westnordost.streetcomplete.osm.opening_hours.model.Weekdays
import de.westnordost.streetcomplete.osm.opening_hours.parser.toOpeningHoursRules
import de.westnordost.streetcomplete.quests.opening_hours.WeekdaysPickerDialog
import de.westnordost.streetcomplete.util.timeOfDayToString
import de.westnordost.streetcomplete.view.dialogs.TimePickerDialog
import kotlinx.serialization.Serializable
import java.util.Locale

@Serializable
data class CollectionTimesRow(var weekdays: Weekdays, var time: Int)

class CollectionTimesAdapter(
    private val context: Context,
    private val countryInfo: CountryInfo
) : RecyclerView.Adapter<CollectionTimesAdapter.ViewHolder>() {

    var collectionTimesRows: MutableList<CollectionTimesRow> = mutableListOf()
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    var isEnabled = true
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    fun createCollectionTimes() = collectionTimesRows.toOpeningHoursRules()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return ViewHolder(QuestTimesWeekdayRowBinding.inflate(inflater, parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val times = collectionTimesRows[position]
        val previousTimes = if (position > 0) collectionTimesRows[position - 1] else null
        holder.update(times, previousTimes, isEnabled)
    }

    override fun getItemCount() = collectionTimesRows.size

    /* ------------------------------------------------------------------------------------------ */

    private fun remove(position: Int) {
        if (!isEnabled) return

        collectionTimesRows.removeAt(position)
        notifyItemRemoved(position)
        // if not last weekday removed -> element after this one may need to be updated
        // because it may need to show the weekdays now
        if (position < collectionTimesRows.size) notifyItemChanged(position)
    }

    fun addNewWeekdays() {
        val isFirst = collectionTimesRows.isEmpty()
        openSetWeekdaysDialog(getWeekdaysSuggestion(isFirst)) { weekdays ->
            openSetTimeDialog(12 * 60) { minutes ->
                add(weekdays, minutes)
            }
        }
    }

    fun addNewHours() {
        val rowAbove = if (collectionTimesRows.size > 0) collectionTimesRows[collectionTimesRows.size - 1] else return
        openSetTimeDialog(12 * 60) { minutes ->
            add(rowAbove.weekdays, minutes)
        }
    }

    private fun add(weekdays: Weekdays, minutes: Int) {
        val insertIndex = itemCount
        collectionTimesRows.add(CollectionTimesRow(weekdays, minutes))
        notifyItemInserted(insertIndex)
    }

    /* ------------------------------------ weekdays select --------------------------------------*/

    inner class ViewHolder(
        private val binding: QuestTimesWeekdayRowBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        init {
            binding.deleteButton.setOnClickListener {
                val index = adapterPosition
                if (index != RecyclerView.NO_POSITION) remove(adapterPosition)
            }
        }

        fun update(times: CollectionTimesRow, previousTimes: CollectionTimesRow?, isEnabled: Boolean) {
            if (previousTimes != null && times.weekdays == previousTimes.weekdays) {
                binding.weekdaysLabel.text = ""
            } else {
                val locale = countryInfo.officialLanguages.firstOrNull()?.let { Locale(it) } ?: Locale.getDefault()
                binding.weekdaysLabel.text = times.weekdays.toLocalizedString(context.resources, locale)
            }

            binding.weekdaysLabel.setOnClickListener {
                openSetWeekdaysDialog(times.weekdays) { weekdays ->
                    times.weekdays = weekdays
                    notifyItemChanged(adapterPosition)
                }
            }
            binding.hoursLabel.text = timeOfDayToString(Locale.getDefault(), times.time)
            binding.hoursLabel.setOnClickListener {
                openSetTimeDialog(times.time) { minutes ->
                    times.time = minutes
                    notifyItemChanged(adapterPosition)
                }
            }

            binding.deleteButton.isInvisible = !isEnabled
            binding.deleteButton.isClickable = isEnabled
            binding.weekdaysLabel.isClickable = isEnabled
            binding.hoursLabel.isClickable = isEnabled
        }
    }

    private fun getWeekdaysSuggestion(isFirst: Boolean): Weekdays {
        if (isFirst) {
            val firstWorkDayIdx = Weekdays.getWeekdayIndex(countryInfo.firstDayOfWorkweek)
            val result = BooleanArray(Weekdays.OSM_ABBR_WEEKDAYS.size)
            for (i in 0 until countryInfo.workweekDays) {
                result[(i + firstWorkDayIdx) % Weekdays.WEEKDAY_COUNT] = true
            }
            return Weekdays(result)
        }
        return Weekdays()
    }

    private fun openSetWeekdaysDialog(weekdays: Weekdays, callback: (weekdays: Weekdays) -> Unit) {
        val locale = countryInfo.officialLanguages.firstOrNull()?.let { Locale(it) } ?: Locale.getDefault()
        WeekdaysPickerDialog.show(context, weekdays, locale, callback)
    }

    private fun openSetTimeDialog(minutes: Int, callback: (minutes: Int) -> Unit) {
        TimePickerDialog(context, minutes / 60, minutes % 60, DateFormat.is24HourFormat(context)) { hourOfDay, minute ->
            callback(hourOfDay * 60 + minute)
        }.show()
    }
}
